File: frontend/src/types/rbac.ts
// Role types for RBAC
export type Role = 'user' | 'admin';

// Permission identifiers matching backend
export type Permission =
  | 'conversations:view'
  | 'conversations:create'
  | 'own_data:manage'
  | 'users:manage'
  | 'organization:manage'
  | 'conversations:view_all'
  | 'providers:manage'
  | 'settings:manage';

// Permission description for UI
export interface PermissionInfo {
  id: Permission;
  name: string;
  description: string;
}

// All available permissions with descriptions
export const PERMISSION_INFO: Record<Permission, PermissionInfo> = {
  'conversations:view': {
    id: 'conversations:view',
    name: 'View Conversations',
    description: 'View your own conversations',
  },
  'conversations:create': {
    id: 'conversations:create',
    name: 'Create Conversations',
    description: 'Create new conversations',
  },
  'own_data:manage': {
    id: 'own_data:manage',
    name: 'Manage Own Data',
    description: 'Manage your own data and settings',
  },
  'users:manage': {
    id: 'users:manage',
    name: 'Manage Users',
    description: 'View and manage all users',
  },
  'organization:manage': {
    id: 'organization:manage',
    name: 'Manage Organization',
    description: 'Manage organization settings',
  },
  'conversations:view_all': {
    id: 'conversations:view_all',
    name: 'View All Conversations',
    description: 'View all conversations in the organization',
  },
  'providers:manage': {
    id: 'providers:manage',
    name: 'Manage Providers',
    description: 'Manage LLM providers and API keys',
  },
  'settings:manage': {
    id: 'settings:manage',
    name: 'Manage Settings',
    description: 'Manage system-wide settings',
  },
};

// Role to permissions mapping
export const ROLE_PERMISSIONS: Record<Role, Permission[]> = {
  user: [
    'conversations:view',
    'conversations:create',
    'own_data:manage',
  ],
  admin: [
    'conversations:view',
    'conversations:create',
    'own_data:manage',
    'users:manage',
    'organization:manage',
    'conversations:view_all',
    'providers:manage',
    'settings:manage',
  ],
};

// Valid roles for validation
export const VALID_ROLES: Role[] = ['user', 'admin'];

// Check if a role string is valid
export function isValidRole(role: string): role is Role {
  return VALID_ROLES.includes(role as Role);
}

// Check if a role has a specific permission
export function hasPermission(role: Role | undefined, permission: Permission): boolean {
  if (!role || !isValidRole(role)) {
    return false;
  }
  return ROLE_PERMISSIONS[role].includes(permission);
}

// Get all permissions for a role
export function getPermissions(role: Role): Permission[] {
  return ROLE_PERMISSIONS[role] || [];
}

// Get permission info for display
export function getPermissionInfo(permission: Permission): PermissionInfo {
  return PERMISSION_INFO[permission];
}

// Role display names
export const ROLE_DISPLAY_NAMES: Record<Role, string> = {
  user: 'User',
  admin: 'Administrator',
};

// Get display name for a role
export function getRoleDisplayName(role: Role): string {
  return ROLE_DISPLAY_NAMES[role] || role;
}

File: backend/internal/security/rbac.go
package security

// Role represents a user's role in the system
type Role string

const (
	// RoleUser is the default role for regular users
	RoleUser Role = "user"
	// RoleAdmin is the role for administrators with elevated privileges
	RoleAdmin Role = "admin"
)

// ValidRoles contains all valid role values
var ValidRoles = []Role{RoleUser, RoleAdmin}

// IsValidRole checks if a role string is a valid role
func IsValidRole(role string) bool {
	for _, r := range ValidRoles {
		if string(r) == role {
			return true
		}
	}
	return false
}

// Permission represents a specific action that can be performed
type Permission string

const (
	// User permissions - available to all authenticated users
	PermViewConversations   Permission = "conversations:view"
	PermCreateConversations Permission = "conversations:create"
	PermManageOwnData       Permission = "own_data:manage"

	// Admin permissions - available only to administrators
	PermManageUsers          Permission = "users:manage"
	PermManageOrganization   Permission = "organization:manage"
	PermViewAllConversations Permission = "conversations:view_all"
	PermManageProviders      Permission = "providers:manage"
	PermManageSettings       Permission = "settings:manage"
)

// ResourceType represents a type of resource in the system
type ResourceType string

const (
	ResourceConversation ResourceType = "conversation"
	ResourceAgent        ResourceType = "agent"
	ResourceWorkflow     ResourceType = "workflow"
	ResourceTool         ResourceType = "tool"
	ResourceUser         ResourceType = "user"
	ResourceOrganization ResourceType = "organization"
	ResourceTeam         ResourceType = "team"
	ResourceSettings     ResourceType = "settings"
	ResourceProvider     ResourceType = "provider"
)

// Action represents an action that can be performed on a resource
type Action string

const (
	ActionView   Action = "view"
	ActionCreate Action = "create"
	ActionUpdate Action = "update"
	ActionDelete Action = "delete"
	ActionExecute Action = "execute"
)

// RolePermissions maps roles to their allowed permissions
var RolePermissions = map[Role][]Permission{
	RoleUser: {
		PermViewConversations,
		PermCreateConversations,
		PermManageOwnData,
	},
	RoleAdmin: {
		// Admins have all user permissions
		PermViewConversations,
		PermCreateConversations,
		PermManageOwnData,
		// Plus admin-specific permissions
		PermManageUsers,
		PermManageOrganization,
		PermViewAllConversations,
		PermManageProviders,
		PermManageSettings,
	},
}

// HasPermission checks if a role has a specific permission
func HasPermission(role Role, perm Permission) bool {
	permissions, ok := RolePermissions[role]
	if !ok {
		return false
	}
	for _, p := range permissions {
		if p == perm {
			return true
		}
	}
	return false
}

// GetPermissions returns all permissions for a role
func GetPermissions(role Role) []Permission {
	permissions, ok := RolePermissions[role]
	if !ok {
		return []Permission{}
	}
	// Return a copy to prevent modification
	result := make([]Permission, len(permissions))
	copy(result, permissions)
	return result
}

// GetPermissionStrings returns all permissions for a role as strings
func GetPermissionStrings(role Role) []string {
	permissions := GetPermissions(role)
	result := make([]string, len(permissions))
	for i, p := range permissions {
		result[i] = string(p)
	}
	return result
}
